File: test/TestCases/napi-dotnet/streams.js
const assert = require('assert');
const fs = require('fs/promises');
const { Duplex, Readable, Writable } = require('stream');

/** @type {import('./napi-dotnet')} */
const binding = require('../common').binding;

const Streams = binding.Streams;
const TestStream = binding.TestStream;

const testFile = 'streams-test.txt';
const testLineCount = 10000;
testStreams();

async function testStreams() {
  await writeToDotNetStreamAsync();
  await readFromDotNetStreamAsync();
  await fs.unlink(testFile);
  await readFromNodeStreamAsync();
  await writeToNodeStreamAsync();
  await readFromTestStreamAsync();
  await writeToTestStreamAsync();
}

/**
 * Gets a Node.js Writable adapter for a .NET stream to the test file, and writes lines to the file.
 */
async function writeToDotNetStreamAsync() {
  const writableStream = Streams.writeFile(testFile);
  assert(writableStream instanceof Writable);

  writableStream.on('error', (err) => {
    console.error(err);
  });

  for (let i = 1; i <= testLineCount; i++) {
    await writeAsync(writableStream, i + '\n');
  }
  await endAsync(writableStream);
  writableStream.destroy();
}

function writeAsync(stream, data) {
  return new Promise((resolve, reject) => {
    const cb = (err) => err ? reject(err) : resolve();
    if (!stream.write(data)) {
      stream.once('drain', cb);
    } else {
      process.nextTick(cb);
    }
  });
}

function endAsync(stream) {
  return new Promise((resolve, reject) => {
    const cb = (err) => err ? reject(err) : resolve();
    stream.end(cb);
  });
}

/**
 * Gets a Node.js Readable adapter for a .NET stream to the test file, reads lines from the file,
 * and validates the expected content.
 */
async function readFromDotNetStreamAsync() {
  const readableStream = Streams.readFile(testFile);
  assert(readableStream instanceof Readable);

  await new Promise((resolve, reject) => {
    let allData = Buffer.alloc(0);
    readableStream.on('error', (err) => {
      console.error(err);
    });
    readableStream.on('data', (chunk) => {
      allData = Buffer.concat([allData, chunk]);
    });
    readableStream.on('end', () => {
      try {
        const lines = allData.toString().trimEnd().split('\n');
        for (let i = 1; i <= testLineCount; i++) {
          assert.equal(lines[i - 1], i.toString());
        }
      } catch (e) { reject(e); }
      resolve();
    });
  });
  readableStream.destroy();
}

/**
 * Creates a Node.js Readable stream and passes it to .NET, which reads from the stream
 * using a .NET stream adapter, then pipes the data to the test file.
 */
async function readFromNodeStreamAsync() {
  const data = Buffer.from(
    [...Array(testLineCount).keys()].map((i) => i + 1).join('\n') + '\n');
  const readableStream = new Readable({
    read() {
      this.push(data);
      this.push(null);
    },
  });

  assert(!readableStream.readableObjectMode);
  await Streams.pipeToFileAsync(testFile, readableStream);
}

/**
 * Creates a Node.js Writable stream that writes to a buffer, then passes the stream to .NET,
 * which pipes data from the test file to the stream using a .NET stream adapter. Then
 * validates the resulting buffer.
 */
async function writeToNodeStreamAsync() {
  let allData = Buffer.alloc(0);
  const writableStream = new Writable({
    write(chunk, encoding, cb) {
      allData = Buffer.concat([allData, chunk]);
      cb();
    },
  });

  await Streams.pipeFromFileAsync(testFile, writableStream);

  const lines = allData.toString().trimEnd().split('\n');
  for (let i = 1; i <= testLineCount; i++) {
    assert.equal(lines[i - 1], i.toString());
  }
}

/**
 * Creates an instance of a .NET Stream subclass and reads from it using a JS adapter.
 */
async function readFromTestStreamAsync()
{
  const data =
    [...Array(testLineCount).keys()].map((i) => i + 1).join('\n') + '\n';

  const readableStream = new TestStream(data);
  assert(readableStream instanceof Readable);

  await new Promise((resolve, reject) => {
    let allData = Buffer.alloc(0);
    readableStream.on('error', (err) => {
      console.error(err);
    });
    readableStream.on('data', (chunk) => {
      allData = Buffer.concat([allData, chunk]);
    });
    readableStream.on('end', () => {
      try {
        const lines = allData.toString().trimEnd().split('\n');
        for (let i = 1; i <= testLineCount; i++) {
          assert.equal(lines[i - 1], i.toString());
        }
      } catch (e) { reject(e); }
      resolve();
    });
  });
  readableStream.destroy();
}

/**
 * Creates an instance of a .NET Stream subclass and writes to it using a JS adapter.
 */
async function writeToTestStreamAsync()
{
  const writableStream = new TestStream(100000);
  assert(writableStream instanceof Writable);

  writableStream.on('error', (err) => {
    console.error(err);
  });

  for (let i = 1; i <= testLineCount; i++) {
    await writeAsync(writableStream, i + '\n');
  }
  await endAsync(writableStream);

  const allData = TestStream.getData(writableStream);
  const lines = allData.trimEnd().split('\n');
  for (let i = 1; i <= testLineCount; i++) {
    assert.equal(lines[i - 1], i.toString());
  }

  writableStream.destroy();
}

File: test/TestCases/napi-dotnet/Streams.cs
using System;
using System.IO;
using System.Text;
using System.Threading.Tasks;

namespace Microsoft.JavaScript.NodeApi.TestCases;

/// <summary>
/// Tests streaming data between .NET Streams and Node.js Duplex streams.
/// </summary>
[JSExport]
public static class Streams
{
    /// <summary>
    /// Gets .NET stream that will be read from by Node.js.
    /// </summary>
    public static Stream ReadFile(string filePath)
    {
        return File.OpenRead(filePath);
    }

    /// <summary>
    /// Gets .NET stream that will be written to by Node.js.
    /// </summary>
    public static Stream WriteFile(string filePath)
    {
        return File.Create(filePath);
    }

    /// <summary>
    /// Reads data from a file and pipes it to a Node.js Writable stream.
    /// </summary>
    public static async Task PipeFromFileAsync(string filePath, Stream toStream)
    {
        using Stream fileStream = File.OpenRead(filePath);

#if NETFRAMEWORK
        byte[] buffer = new byte[4096];
        int count;
        while ((count = await fileStream.ReadAsync(buffer, 0, buffer.Length)) > 0)
        {
            await toStream.WriteAsync(buffer, 0, count);
        }
#else
        Memory<byte> buffer = new byte[4096].AsMemory();
        int count;
        while ((count = await fileStream.ReadAsync(buffer)) > 0)
        {
            await toStream.WriteAsync(buffer.Slice(0, count));
        }
#endif
    }

    /// <summary>
    /// Creates a file and pipes data from a Node.js Readable stream to the file.
    /// </summary>
    public static async Task PipeToFileAsync(string filePath, Stream fromStream)
    {
        using Stream fileStream = File.Create(filePath);

#if NETFRAMEWORK
        byte[] buffer = new byte[4096];
        int count;
        while ((count = await fromStream.ReadAsync(buffer, 0, buffer.Length)) > 0)
        {
            await fileStream.WriteAsync(buffer, 0, count);
        }
#else
        Memory<byte> buffer = new byte[4096].AsMemory();
        int count;
        while ((count = await fromStream.ReadAsync(buffer)) > 0)
        {
            await fileStream.WriteAsync(buffer.Slice(0, count));
        }
#endif
    }
}

[JSExport]
public class TestStream : Stream
{
    private readonly MemoryStream _memory;

    public TestStream(string data)
    {
        _memory = new MemoryStream(Encoding.UTF8.GetBytes(data));
    }

    public TestStream(int capacity)
    {
        _memory = new MemoryStream(capacity);
    }

    public static string GetData(TestStream stream)
    {
        return Encoding.UTF8.GetString(stream._memory.ToArray());
    }

    public override bool CanRead => true;
    public override bool CanWrite => true;
    public override bool CanSeek => false;

    public override long Length => _memory.Length;

    public override long Position
    {
        get => _memory.Position;
        set => _memory.Position = value;
    }

    public override void SetLength(long value) => _memory.SetLength(value);

    public override int Read(byte[] buffer, int offset, int count)
        => _memory.Read(buffer, offset, count);

    public override void Write(byte[] buffer, int offset, int count)
        => _memory.Write(buffer, offset, count);

    public override long Seek(long offset, SeekOrigin origin)
        => _memory.Seek(offset, origin);

    public override void Flush() => _memory.Flush();
}
